File: Android/app/src/main/java/catglo/com/api/GoogleAddressSuggester.java
package catglo.com.api;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.location.Criteria;
import android.location.Location;
import android.location.LocationManager;
import android.preference.PreferenceManager;

import catglo.com.deliveryDatabase.AddressInfo;
import catglo.com.deliverydroid.BuildConfig;
import catglo.com.deliverydroid.data.MyGeoPoint;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;


import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.ArrayList;

/**
 * Created by goblets on 2/27/14.
 */
public class GoogleAddressSuggester {

    private final SharedPreferences sharedPreferences;
    private final String addressFilterComponents;
    private final RequestQueue requestQue;
    protected final Context context;

    public interface AddressListListener {
        void commit(ArrayList<AddressInfo> addressList, String searchString);
    }

    public AddressListListener resultListener;

    JsonObjectRequest request;
    Response.ErrorListener errorListener = new Response.ErrorListener() {
        @Override
        public void onErrorResponse(VolleyError error) {
        //Log.e("REST", error.getLocalizedMessage());
        }
    };

    public GoogleAddressSuggester(Context context, AddressListListener addressListListener) {

        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        addressFilterComponents = sharedPreferences.getString("addressFilterComponents", "");

        this.resultListener = addressListListener;
        requestQue = Volley.newRequestQueue(context);
        this.context = context;
    }

    public void lookup(final String lookupAddress){
        String url;
        String address = null;
        try {
            address = URLEncoder.encode(lookupAddress, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
            address = "";
        }

        float range = sharedPreferences.getFloat("dileveryRadius", 1f); //TODO: Make configurable!!

        LocationManager locationManager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
        Criteria criteria = new Criteria();
        String bestProvider = locationManager.getBestProvider(criteria, false);

        @SuppressLint("MissingPermission") Location location = locationManager.getLastKnownLocation(bestProvider);
        if (location != null) {//TODO: MAke a better solution, gps from prefs..
            double lat = location.getLatitude();
            double lng = location.getLongitude();
            String bounds = "&bounds=" + (lat - range) + "," + (lng - range) + URLEncoder.encode("|") + (lat + range) + "," + (lng + range);
            url = "https://maps.googleapis.com/maps/api/geocode/json?address=" + address + bounds + "&sensor=true&key="+BuildConfig.GoogleMapsApiKey;
        } else {
            url = "https://maps.googleapis.com/maps/api/geocode/json?address=" + address + "&sensor=true&key="+BuildConfig.GoogleMapsApiKey;
        }

        if (addressFilterComponents.length() > 1) {
            url = url + "&components=" + URLEncoder.encode(addressFilterComponents);
        }

        requestQue.add(new JsonObjectRequest(url, null, new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject jsonResponse) {

                try {
                    JSONArray results = jsonResponse.getJSONArray("results");
                    ArrayList<AddressInfo> addresses = new ArrayList<AddressInfo>();

                    for (int index = 0; index < results.length(); index++){
                        JSONObject a1=results.getJSONObject(index);

                        AddressInfo addressInfo = new AddressInfo();
                        addressInfo.address = a1.getString("formatted_address");

                        boolean includedRoute;
                        includedRoute=false;
                        JSONArray address_components = a1.getJSONArray("address_components");
                        for (int i = 0; i < address_components.length(); i++){
                            JSONArray types = address_components.getJSONObject(i).getJSONArray("types");
                            for (int j = 0; j < types.length(); j++){
                                if (types.getString(j).equalsIgnoreCase("route")){
                                    includedRoute=true;
                                }
                            }
                        }
                        if (includedRoute){
                            try {
                                JSONObject geometry = a1.getJSONObject("geometry");
                                JSONObject location = geometry.getJSONObject("location");
                                float lat = Float.parseFloat(location.getString("lat"));
                                float lng = Float.parseFloat(location.getString("lng"));
                                addressInfo.location = new MyGeoPoint((int)(lat*1e6),(int)(lng*1e6));
                            } catch (NumberFormatException e){};

                            addresses.add(addressInfo);
                        }
                    }

                    resultListener.commit(addresses, lookupAddress);
                } catch (JSONException e) {
                    e.printStackTrace();
                }

            }
        }, errorListener));

    }

}

File: Android/app/src/main/java/catglo/com/api/GoogleAddressSuggester.kt
package catglo.com.api


import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.location.Criteria
import android.location.Geocoder
import android.location.LocationManager
import android.preference.PreferenceManager
import catglo.com.deliveryDatabase.AddressInfo
import com.google.android.gms.location.places.*
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import com.google.android.gms.tasks.Task
import java.util.*


/**
 * Created by goblets on 2/27/14.
 */
open class GoogleAddressSuggester(protected val context: Context, var resultListener: AddressListListener) {

    private var sharedPreferences: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)

    interface AddressListListener {
        fun commit(addressList: ArrayList<AddressInfo>, searchString: String)
    }

    private var geocoder: Geocoder = Geocoder(context)
    private var geoDataClient: GeoDataClient = Places.getGeoDataClient(context)
    private var placesDetectionClient : PlaceDetectionClient = Places.getPlaceDetectionClient(context)
//    GeoDataClient.getAutocompletePredictions()

    internal inner class AutocompletePredictor(var query: String) {

        fun getPredictions(
            bounds: LatLngBounds,
            typeFilter: AutocompleteFilter
        ): Task<AutocompletePredictionBufferResponse> {
            return geoDataClient.getAutocompletePredictions(query, bounds, typeFilter)
        }
    }


    private var lastQuery: String = ""

    @SuppressLint("MissingPermission")
    open fun lookup(searchAddress: String) {

        val range = 0.2f //sharedPreferences.getFloat("dileveryRadius", 1f) //TODO: Make configurable!!


        val addressInfoList = ArrayList<AddressInfo>()
        (context.getSystemService(Context.LOCATION_SERVICE) as LocationManager).run {
            val criteria = Criteria()
            val bestProvider = getBestProvider(criteria, false)
            val location = getLastKnownLocation(bestProvider)


            if (location != null) {
                var bounds = LatLngBounds(LatLng((location.latitude - range),(location.longitude - range)),
                                          LatLng((location.latitude + range),(location.longitude + range)))

                lastQuery = searchAddress

                // Submit the query to the autocomplete API and retrieve a PendingResult that will contain the results when the query completes.
                val typeFilter = AutocompleteFilter.Builder().setTypeFilter( AutocompleteFilter.TYPE_FILTER_ADDRESS or AutocompleteFilter.TYPE_FILTER_GEOCODE).build()
                val predictor = AutocompletePredictor(searchAddress)
                val results = predictor.getPredictions(bounds, typeFilter)

                results.addOnSuccessListener { autocompletePredictions ->
                    if (resultListener != null /*&& predictor.query == lastQuery*/) {
                        autocompletePredictions.forEach {prediction ->
                            val addressString = prediction.getFullText(null)
                            addressInfoList.add(AddressInfo(addressString.toString()))
                        }
                        resultListener.commit(addressInfoList, searchAddress)
                    }
                    autocompletePredictions.release()
                }
            }
        }
    }

}
